### ISSUE TITLE
A ResourcePool cannot acquire a new resource -- the factory or source appears to be down.

### ISSUE THREAD

**Issue by winterth** [swaldman/c3p0#117]
Hi,
In the method `forceKillAcquires` in `BasicResourcePool`, it sets the `force_kill_acquires` before `wait()`. And this flag would be clear after `wait()`. But if the `wait()` was interrupted, the `force_kill_acquires` will always be true.

Then, every getConnection would got this exception `A ResourcePool cannot acquire a new resource -- the factory or source appears to be down.`

Can `force_kill_acquires` be moved into the finally?

```
    private synchronized void forceKillAcquires() throws InterruptedException
    {
	if (logger.isLoggable(MLevel.WARNING))
	    logger.log(MLevel.WARNING,
		       "Having failed to acquire a resource, " +
		       this +
		       " is interrupting all Threads waiting on a resource to check out. " +
		       "Will try again in response to new client requests.");

        Thread t = Thread.currentThread();

        try
        {
            force_kill_acquires = true;
            this.notifyAll(); //wake up any threads waiting on an acquire, and force them all to die.
            while (acquireWaiters.size() > 0) //we want to let all the waiting acquires die before we unset force_kill_acquires
            {
                otherWaiters.add( t );
                this.wait();
            }
            force_kill_acquires = false;
        }
        finally
        { otherWaiters.remove( t ); }
    }
```

```
    private void awaitAvailable(long timeout) throws InterruptedException, TimeoutException, ResourcePoolException
    {
        assert Thread.holdsLock( this );

        if (force_kill_acquires)
            throw new ResourcePoolException("A ResourcePool cannot acquire a new resource -- the factory or source appears to be down.");

        Thread t = Thread.currentThread();
        try
        {
```

**Comment by gonfi** [comment]
I run into the same issue that is described here https://sourceforge.net/p/c3p0/bugs/130/ and now searched the list of issues in this GitHub repo. Maybe this suggested change would fix the issue?

My scenario was the same: DB connection lost for a prolonged time, then came back, but my application that uses c3p0 (and Spring) could not recover, ever. This is sad.

I'm using a very old c3p0: 0.9.2.1 from 2013, but somehow feel that it makes no difference.

Restarting my Java app solved the problem for now. I'll try to upgrade to 0.9.5.3 from 2019 but won't try to reproduce the problem.

**Comment by swaldman** [comment]
I think this issue was addressed as of c3p0-0.9.5.4, see https://github.com/swaldman/c3p0/commit/5a9b0ebc331e3a3e6f512cd551aed25c19f0beef

Let me know if you are still seeing anything like it.

Thanks!
